ec/src/wnaf.rs

```diff
@@ -0,0 +1,77 @@
+use crate::ProjectiveCurve;
+use ark_ff::{BigInteger, PrimeField};
+use ark_std::vec::Vec;
+
+/// A helper type that contains all the context required for computing
+/// a window NAF multiplication of a group element by a scalar.
+pub struct WnafContext {
+    pub window_size: usize,
+}
+
+impl WnafContext {
+    /// Construct a new context for a window of size `window_size`.
+    pub fn new(window_size: usize) -> Self {
+        assert!(window_size >= 2);
+        assert!(window_size < 64);
+        Self { window_size }
+    }
+
+    pub fn table<G: ProjectiveCurve>(&self, mut base: G) -> Vec<G> {
+        let mut table = Vec::with_capacity(1 << (self.window_size - 1));
+        let dbl = base.double();
+
+        for _ in 0..(1 << (self.window_size - 1)) {
+            table.push(base);
+            base += &dbl;
+        }
+        table
+    }
+
+    /// Computes scalar multiplication of a group element `g` by `scalar`.
+    ///
+    /// This method uses the wNAF algorithm to perform the scalar multiplication;
+    /// first, it uses `Self::table` to calculate an appropriate table of multiples of `g`,
+    /// and then uses the wNAF algorithm to compute the scalar multiple.
+    pub fn mul<G: ProjectiveCurve>(&self, g: G, scalar: &G::ScalarField) -> G {
+        let table = self.table(g);
+        self.mul_with_table(&table, scalar).unwrap()
+    }
+
+    /// Computes scalar multiplication of a group element by `scalar`.
+    /// `base_table` holds precomputed multiples of the group element; it can be generated using `Self::table`.
+    /// `scalar` is an element of `G::ScalarField`.
+    ///
+    /// Returns `None` if the table is too small.
+    pub fn mul_with_table<G: ProjectiveCurve>(
+        &self,
+        base_table: &[G],
+        scalar: &G::ScalarField,
+    ) -> Option<G> {
```

Comment:
@ValarDragon @jon-chuang should we normalize the table here?

Also, should we make `fixed_base_msm` use wNAF also?

Reply:
I think normalizing the table could be done only for fixedbaseMSM? 

Unless mixed addition is significantly faster than the ordinary projective add? Still, for fixed base MSM, one can potentially use the infrastructure created for the batch ops. 

Is it used for something important? I can't recall.

Reply:
Yeah, I was proposing using mixed addition instead of projective add. It's not always much faster, for example it's not worth it for Edwards curves. Another thing that we can do is to make two different methods, one that uses a projective table, and one that use an affine table. (Maybe we can abstract over these?)

Reply:
Let's revisit this later; there's no need to block this PR on this decision.